internal/timex/timextest/fake_clock.go

```diff
@@ -0,0 +1,168 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+// Package timextest provides fake controlled time for testing code using the [timex] package. All methods here are safe
+// to call concurrently.
+package timextest
+
+import (
+	"sync"
+	"time"
+
+	"go.temporal.io/server/internal/timex"
+)
+
+type (
+	// FakeClock is a timex.Clock. Unlike other fake clock implementations, the methods of this clock are synchronous, so
+	// when you call Advance or Update, all triggered timers from AfterFunc will fire before the method returns, in the same
+	// goroutine.
+	FakeClock struct {
+		mu     sync.RWMutex
+		now    time.Time
+		timers []*fakeTimer
+	}
+
+	// fakeTimer is a fake implementation of [timex.Timer].
+	fakeTimer struct {
+		// link to the parent fake time source
+		clock *FakeClock
+		// deadline for when the timer should fire
+		deadline time.Time
+		// callback to call when the timer fires
+		callback func()
+		// called indicates if the timer has already count
+		called bool
+		// index of the timer in the parent fake time source
+		index int
+	}
+)
+
+// NewFakeClock returns a FakeClock with the current time set to the zero value of [time.Time].
+func NewFakeClock() *FakeClock {
+	return &FakeClock{
+		now: time.Time{},
+	}
+}
+
+// Now return the current time.
+func (c *FakeClock) Now() time.Time {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	return c.now
+}
+
+// AfterFunc return a timer that will fire after the specified duration. It is important to note that the clock is
+// locked while the callback is called. This means that you must be cautious about calling any methods on the clock from
+// within the callback. Doing so will probably result in a deadlock. To avoid this, you may want to wrap all such calls
+// in a goroutine.
+func (c *FakeClock) AfterFunc(d time.Duration, f func()) timex.Timer {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	t := &fakeTimer{clock: c, deadline: c.now.Add(d), callback: f}
+	t.index = len(c.timers)
+	c.timers = append(c.timers, t)
+
+	return t
+}
+
+// Update the fake current time.
+func (c *FakeClock) Update(now time.Time) *FakeClock {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	return c.update(now)
+}
+
+// Advance the timer by the specified duration.
+func (c *FakeClock) Advance(d time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.update(c.now.Add(d))
+}
+
+// update the fake current time without locking.
+func (c *FakeClock) update(now time.Time) *FakeClock {
+	c.now = now
+	c.fireTimers()
+
+	return c
+}
+
+// fireTimers fires all timers that are ready.
+func (c *FakeClock) fireTimers() {
+	for i := len(c.timers) - 1; i >= 0; i-- {
+		if c.timers[i].deadline.After(c.now) {
+			continue
+		}
+
+		c.timers[i].callback()
+		c.removeTimer(c.timers[i])
+	}
```

Comment:
It needs to be in reverse for removeTimer to work since it truncates the list. I think your approach is definitely better, though. I can't commit it as-is because I need to update the timer index as well so that calls to `.Stop()` will work, but I'll make an update with this one.

Reply:
Why do you need a timer index? (can discuss in later comment)

Reply:
Yeah, it's not necessary, but it's more efficient, and IMO it's worth the tradeoff.